### PR TITLE
Prediction engine for time series.

### PR DESCRIPTION
**Pull request by codemzs**
Prediction engine for time series that updates the time series transform state at the time of prediction. Next iteration will add more tests.

### CODE REVIEW

File: src/Microsoft.ML.Core/Data/MetadataUtils.cs [hunk 1]
```diff
@@ -82,6 +82,8 @@ public static class Kinds
             /// slots within that column.
             /// </summary>
             public const string CategoricalSlotRanges = "CategoricalSlotRanges";
+
+            public const string TimeSeriesColumn = "TimeSeriesColumn";
```
Comment (TomFinley) [inline src/Microsoft.ML.Core/Data/MetadataUtils.cs]: >TimeSeriesColumn [](start = 32, length = 16)

Why is it necessary to add this?

File: src/Microsoft.ML.Data/DataView/CompositeRowToRowMapper.cs [hunk 2]
```diff
@@ -18,6 +19,8 @@ public sealed class CompositeRowToRowMapper : IRowToRowMapper
         public Schema InputSchema { get; }
         public Schema Schema { get; }
 
+        public IRowToRowMapper[] InnerMappers => _innerMappers;
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/DataView/CompositeRowToRowMapper.cs]: >public IRowToRowMapper[] InnerMappers => _innerMappers; [](start = 8, length = 55)

Figure out another way.

File: src/Microsoft.ML.TimeSeries/PredictionEngine.cs [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+{
+    /// <summary>
+    /// A class that runs the previously trained model (and the preceding transform pipeline) on the
+    /// in-memory data, one example at a time.
+    /// This can also be used with trained pipelines that do not end with a predictor: in this case, the
+    /// 'prediction' will be just the outcome of all the transformations.
+    /// </summary>
+    /// <typeparam name="TSrc">The user-defined type that holds the example.</typeparam>
+    /// <typeparam name="TDst">The user-defined type that holds the prediction.</typeparam>
+    public sealed class PredictionEngine<TSrc, TDst>
```
Comment (TomFinley) [inline src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: >PredictionEngine [](start = 24, length = 16)

This appears to have the same name, and indeed 90% of the same code, as the existing prediction engine. The same name is bad, because that's a name collision between two publicly facing types in our API (albeit in different namespaces). We should avoid that.

Also I feel compelled to point out that there are ways for classes to share code beyond copy and paste.

File: test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs [hunk 4]
```diff
@@ -109,19 +120,62 @@ public void ChangePointDetectionWithSeasonality()
...
+                for (int i = 0; i < SeasonalitySize; i++)
+                    data.Add(new Data(i));
+
+                for (int i = 0; i < ChangeHistorySize; i++)
+                    data.Add(new Data(i * 100));
+
+                // Train
+                SsaChangePointDetector detector = new SsaChangePointEstimator(env, args).Fit(dataView);
+                var engine = detector.MakeTimeSeriesPredictionFunction<Data, Prediction1>(env);
+                //Even though time series column is not requested it will pass the observation through time series transform.
```
Comment (Zruty0) [inline test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs]: >//Even though time series column is not requested it will pass the observation through time series transform. [](start = 16, length = 109)

I wonder if this can be somehow checked. Probably not until you have saving implemented

File: src/Microsoft.ML.TimeSeries/PredictionEngine.cs [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+            for (int i = 0; i < innerMappers.Length; ++i)
+            {
+                Action localDisp;
+                if (innerMappers[i] is CompositeRowToRowMapper)
+                    GetStatefulRows(result, innerMappers[i], deps[i], rows, out localDisp, out result);
+                else
+                    result = innerMappers[i].GetRow(result, deps[i], out localDisp);
+
+                if (result is IStatefulRow)
+                    rows.Add((IStatefulRow)input);
```
Comment (Zruty0) [inline src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: >input [](start = 43, length = 5)

`result`, not input.

I think this recursive function is not currently tested, and I think there's more bugs here. Please write a test that exercises this code

File: src/Microsoft.ML.TimeSeries/PredictionEngine.cs [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+            List<IStatefulRow> rows = new List<IStatefulRow>();
+            if (mapper is CompositeRowToRowMapper)
+                GetStatefulRows(_inputRow, mapper, col => true, rows, out _disposer, out var outRow);
+
+            var cursorable = TypedCursorable<TDst>.Create(env, new EmptyDataView(env, mapper.Schema), ignoreMissingColumns, outputSchemaDefinition);
+            var outputRow = mapper.GetRow(_inputRow, col => true, out _disposer);
+
+            if (rows.Count == 0 && outputRow is IStatefulRow)
+                rows.Add((IStatefulRow)outputRow);
+
```
Comment (Zruty0) [inline src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: this feels like an unroll of 1 level of recursion. Can't `GetStatefulRows` also handle case where mapper is not a `CompositeRowToRowMapper`?

File: src/Microsoft.ML.TimeSeries/PredictionEngine.cs (hunk 5 above)
Comment (codemzs) [reply src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: Yep, I realized that when I added another test an hour back but thanks for also looking at this. In the next iteration this will be fixed.

File: src/Microsoft.ML.Api/PredictionEngine.cs [hunk 7]
```diff
@@ -132,20 +155,22 @@ public void Reset()
...
         where TSrc : class
         where TDst : class, new()
     {
         private readonly DataViewConstructionUtils.InputRow<TSrc> _inputRow;
         private readonly IRowReadableAs<TDst> _outputRow;
         private readonly Action _disposer;
 
-        internal PredictionEngine(IHostEnvironment env, Stream modelStream, bool ignoreMissingColumns,
+        [BestFriend]
+        internal PredictionEngineBase(IHostEnvironment env, Stream modelStream, bool ignoreMissingColumns,
```
Comment (TomFinley) [inline src/Microsoft.ML.Api/PredictionEngine.cs]: Incidentally, given that this is a base class, there is a new concept in either C# 7.2 or 7.3 or something called "private protected" which is precisely for this scenario, perhaps you could use it here, now that this is a base class.

File: src/Microsoft.ML.TimeSeries/PredictionEngine.cs [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+                    else
+                        disposer = localDisp + disposer;
+                    // We want the last disposer to be called first, so the order of the addition here is important.
+                }
+            }
+
+            outRow = result;
+        }
+
+        private ValuePinger[][] CreatePingers(List<IStatefulRow> rows)
```
Comment (TomFinley) [inline src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: Pinger I'm not too wild about. How about "tickler?"

Comment (TomFinley) [reply src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: Or maybe we can have the verb be tickle and the thing that does the tickling called feather.

File: src/Microsoft.ML.Api/PredictionEngine.cs [hunk 9]
```diff
@@ -124,6 +124,29 @@ public void Reset()
...
+        {
+        }
+
+        internal PredictionEngine(IHostEnvironment env, IDataView dataPipe, bool ignoreMissingColumns,
+            SchemaDefinition inputSchemaDefinition = null, SchemaDefinition outputSchemaDefinition = null)
+            : base(env, dataPipe, ignoreMissingColumns, inputSchemaDefinition, outputSchemaDefinition)
+        {
+        }
+
+        internal PredictionEngine(IHostEnvironment env, ITransformer transformer, bool ignoreMissingColumns,
```
Comment (Zruty0) [inline src/Microsoft.ML.Api/PredictionEngine.cs]: >PredictionEngine [](start = 17, length = 16)

only keep this constructor and remove the other 2. I think you need to merge with latest master, because these other 2 ctors have been removed from `PredictionEngine`, IIRC

File: src/Microsoft.ML.Api/PredictionEngine.cs [hunk 10]
```diff
@@ -132,20 +155,22 @@ public void Reset()
...
 
-        internal PredictionEngine(IHostEnvironment env, Stream modelStream, bool ignoreMissingColumns,
+        [BestFriend]
+        internal PredictionEngineBase(IHostEnvironment env, Stream modelStream, bool ignoreMissingColumns,
             SchemaDefinition inputSchemaDefinition = null, SchemaDefinition outputSchemaDefinition = null)
             : this(env, StreamChecker(env, modelStream), ignoreMissingColumns, inputSchemaDefinition, outputSchemaDefinition)
         {
         }
 
+        [BestFriend]
```
Comment (Zruty0) [inline src/Microsoft.ML.Api/PredictionEngine.cs]: >BestFriend [](start = 9, length = 10)

this doesn't help at all. In fact, please remove `StreamChecker` altogether, since it's not being used

File: src/Microsoft.ML.Api/PredictionEngine.cs [hunk 11]
```diff
@@ -207,19 +239,24 @@ public TDst Predict(TSrc example)
             return result;
         }
 
+        public void ExtractValues(TSrc example) => _inputRow.ExtractValues(example);
```
Comment (Zruty0) [inline src/Microsoft.ML.Api/PredictionEngine.cs]: >public [](start = 8, length = 6)

that's weird, why do you need it public?

File: src/Microsoft.ML.TimeSeries/PredictionEngine.cs [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.ML.TimeSeries
+{
+    public delegate void ValuePinger(ref bool value, int rowPosition);
+
+    public interface IStatefulRow : IRow
+    {
+        ValuePinger[] GetPingers();
```
Comment (Zruty0) [inline src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: >ValuePinger[] [](start = 8, length = 13)

1. Why do you need value (and why of type bool)? 
2. Why this is an array?

Comment (Zruty0) [reply src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: I believe you can get away with just `Action<long> GetPinger`

---
In reply to: [235552903](https://github.com/dotnet/machinelearning/pull/1618#discussion_r235552903) [](ancestors = 235552903)

File: src/Microsoft.ML.TimeSeries/PredictionEngine.cs [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    public interface IStatefulRowMapper : IRowMapper
+    {
+        Delegate[] CreatePingers(IRow input, Func<int, bool> activeOutput, out Action disposer);
```
Comment (Zruty0) [inline src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: >Delegate[] CreatePingers(IRow input, Func<int, bool> activeOutput, out Action disposer); [](start = 8, length = 88)

why do you need to have this method? I think, regardless of anything, you must call the pinger

File: src/Microsoft.ML.TimeSeries/PredictionEngine.cs (hunk 12 above)
Comment (Zruty0) [reply src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: Or maybe `void Ping(long position)`

---
In reply to: [235552952](https://github.com/dotnet/machinelearning/pull/1618#discussion_r235552952) [](ancestors = 235552952,235552903)

Comment (Zruty0) [reply src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: Or `void Ping()`

---
In reply to: [235553822](https://github.com/dotnet/machinelearning/pull/1618#discussion_r235553822) [](ancestors = 235553822,235552952,235552903)

File: src/Microsoft.ML.TimeSeries/PredictionEngine.cs [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+                        disposer = localDisp + disposer;
+                    // We want the last disposer to be called first, so the order of the addition here is important.
+                }
+            }
+
+            outRow = result;
+        }
+
+        private ValuePinger[][] CreatePingers(List<IStatefulRow> rows)
+        {
```
Comment (Zruty0) [inline src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: >{ [](start = 8, length = 1)

I think you can just merge them all onto one delegate, there's no need to have arrays of arrays of delegates

File: src/Microsoft.ML.TimeSeries/PredictionEngine.cs [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+        /// <param name="prediction">The object to store the prediction in. If it's <c>null</c>, a new one will be created, otherwise the old one
+        /// is reused.</param>
+        public override void Predict(TSrc example, ref TDst prediction)
+        {
+            Contracts.CheckValue(example, nameof(example));
+            ExtractValues(example);
+            if (prediction == null)
+                prediction = new TDst();
+
+            //Update State.
```
Comment (Zruty0) [inline src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: >//Up [](start = 12, length = 4)

`// Update state.` and `// Predict.`

File: src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs [hunk 16]
```diff
@@ -407,4 +524,320 @@ public ValueGetter<TValue> GetGetter<TValue>(int col)
...
         }
     }
+
+    /// <summary>
+    /// This class is a transform that can add any number of output columns, that depend on any number of input columns.
+    /// It does so with the help of an <see cref="IRowMapper"/>, that is given a schema in its constructor, and has methods
+    /// to get the dependencies on input columns and the getters for the output columns, given an active set of output columns.
+    /// </summary>
+
+    public sealed class TimeSeriesRowToRowMapperTransform : RowToRowTransformBase, IRowToRowMapper,
```
Comment (Zruty0) [inline src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs]: >TimeSeriesRowToRowMapperTransform [](start = 24, length = 33)

Why did you need to clone it, and why is it public?

File: test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs (hunk 4 above)
Comment (codemzs) [reply test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs]: I added this check in another test.

---
In reply to: [233639726](https://github.com/dotnet/machinelearning/pull/1618#discussion_r233639726) [](ancestors = 233639726)

File: src/Microsoft.ML.Api/PredictionEngine.cs [hunk 17]
```diff
@@ -189,19 +211,24 @@ public TDst Predict(TSrc example)
...
+        public void FillValues(TDst prediction) => _outputRow.FillValues(prediction);
+
         /// <summary>
         /// Run prediction pipeline on one example.
         /// </summary>
         /// <param name="example">The example to run on.</param>
         /// <param name="prediction">The object to store the prediction in. If it's <c>null</c>, a new one will be created, otherwise the old one
         /// is reused.</param>
-        public void Predict(TSrc example, ref TDst prediction)
+        public virtual void Predict(TSrc example, ref TDst prediction)
```
Comment (Zruty0) [inline src/Microsoft.ML.Api/PredictionEngine.cs]: >virtual [](start = 15, length = 7)

If 2 existing implementations are dissimilar, what's the value in having a virtual method? Make it `abstract`

File: src/Microsoft.ML.Api/PredictionFunction.cs [hunk 18]
```diff
@@ -7,30 +7,40 @@
 
 namespace Microsoft.ML.Runtime.Data
 {
+    public sealed class PredictionFunction<TSrc, TDst> : PredictionFunctionBase<TSrc, TDst>
```
Comment (Zruty0) [inline src/Microsoft.ML.Api/PredictionFunction.cs]: >PredictionFunction [](start = 24, length = 18)

I honestly don't understand the purpose of `PredictionFunction`. Why introduce a hierarchy of useless classes? I would much rather rename `PredictionEngine` to `PredictionFunction` and be done with it.

It doesn't have to be done in this PR, but could you just use `PredictionEngine` only, to simplify that future change?

File: src/Microsoft.ML.Api/PredictionFunction.cs [hunk 19]
```diff
@@ -7,30 +7,40 @@
...
         {
             Contracts.CheckValue(env, nameof(env));
             env.CheckValue(transformer, nameof(transformer));
 
             IDataView dv = env.CreateDataView(new TSrc[0]);
-            _engine = env.CreatePredictionEngine<TSrc, TDst>(transformer);
+            CreatePredictionEngine(env, transformer, out _engine);
         }
 
+        internal virtual void CreatePredictionEngine(IHostEnvironment env, ITransformer transformer, out PredictionEngineBase<TSrc, TDst> engine) =>
```
Comment (Zruty0) [inline src/Microsoft.ML.Api/PredictionFunction.cs]: >CreatePredictionEngine [](start = 30, length = 22)

use return value instead of `out` parameter

File: src/Microsoft.ML.Data/DataLoadSave/TransformerChain.cs [hunk 20]
```diff
@@ -36,20 +36,33 @@ public enum TransformerScope
...
+        TransformerScope[] Scopes { get; }
+    }
+
     /// <summary>
     /// A chain of transformers (possibly empty) that end with a <typeparamref name="TLastTransformer"/>.
     /// For an empty chain, <typeparamref name="TLastTransformer"/> is always <see cref="ITransformer"/>.
     /// </summary>
-    public sealed class TransformerChain<TLastTransformer> : ITransformer, ICanSaveModel, IEnumerable<ITransformer>
+    public sealed class TransformerChain<TLastTransformer> : ITransformer, ICanSaveModel, IEnumerable<ITransformer>, ITransformerAccessor
     where TLastTransformer : class, ITransformer
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/DataLoadSave/TransformerChain.cs]: This is confusing. You now can access the inner transformers in 3 ways: via `foreach` (as an `IEnumerable`), via the internal field `Transformers` and via `ITransformerAccessor.Transformers`. Let's reduce it to one. Maybe make in an `IEnumerable` of pairs, if you need access to `scopes`

Comment (Zruty0) [reply src/Microsoft.ML.Data/DataLoadSave/TransformerChain.cs]: Or you could make `Transformers` and `Scopes` public and `IReadOnlyList`s

---
In reply to: [236007659](https://github.com/dotnet/machinelearning/pull/1618#discussion_r236007659) [](ancestors = 236007659)

File: src/Microsoft.ML.TimeSeries/SequentialAnomalyDetectionTransformBase.cs [hunk 21]
```diff
@@ -306,10 +307,10 @@ public abstract class AnomalyDetectionStateBase : StateBase
             protected SequentialAnomalyDetectionTransformBase<TInput, TState> Parent;
 
             // A windowed buffer to cache the update values to the martingale score in the log scale.
-            private FixedSizeQueue<Double> _logMartingaleUpdateBuffer;
+            private FixedSizeQueue<Double> LogMartingaleUpdateBuffer { get; set; }
```
Comment (Zruty0) [inline src/Microsoft.ML.TimeSeries/SequentialAnomalyDetectionTransformBase.cs]: >LogMartingaleUpdateBuffer [](start = 43, length = 25)

why this change?

**Comment by Zruty0** [comment]
> using Microsoft.ML.Core.Data;

summary comment

---
Refers to: src/Microsoft.ML.TimeSeries/PredictionEngine.cs:1 in 7bf4c39. [](commit_id = 7bf4c394093ec5fe8f4b7ba981a2a8e537d89577, deletion_comment = False)

File: src/Microsoft.ML.TimeSeries/SequentialAnomalyDetectionTransformBase.cs (hunk 21 above)
Comment (Zruty0) [reply src/Microsoft.ML.TimeSeries/SequentialAnomalyDetectionTransformBase.cs]: also, could you add this assembly to be tracked by RoboTom?

---
In reply to: [236008010](https://github.com/dotnet/machinelearning/pull/1618#discussion_r236008010) [](ancestors = 236008010)

File: src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs [hunk 22]
```diff
@@ -191,7 +256,9 @@ public void ProcessWithoutBuffer(ref TInput input, ref TOutput output)
         internal readonly string OutputColumnName;
         private protected ColumnType OutputColumnType;
 
-        public bool IsRowToRowMapper => false;
+        public bool IsRowToRowMapper => true;
+
+        public TState StateRef{ get; set; }
```
Comment (Zruty0) [inline src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs]: >public TState StateRef{ get; set; } [](start = 8, length = 35)

I don't see this as a good idea at all.
The whole cloning mechanism could be implemented completely internally, without making anything public, or mutable. Just don't use `MemberwiseClone`, and write it yourself. This also has the benefit of self-documenting what's being cloned and how.

**Comment by Zruty0** [comment]
you'll need to revert your changes to `libmf`.

File: src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs [hunk 23]
```diff
@@ -190,7 +227,7 @@ public void ProcessWithoutBuffer(ref TInput input, ref TOutput output)
         internal readonly string OutputColumnName;
         private protected ColumnType OutputColumnType;
 
-        public bool IsRowToRowMapper => false;
+        public bool IsRowToRowMapper => true;
```
[the remark this answers is not in the document]
Comment (codemzs) [reply src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs]: In reality TS is a rowtorowmapper but we will just pretend it is not so that it does not work in the other prediction engine. In TS we will recognize this is actually rowtorowmapper covered by a custom interface that extends IRowToRowMapper. Then we go over transformers one by one in the chain and call GetRow on all of them to get the final transformer. How is this supposed to invalidate check-pointing or cloning? 

---
In reply to: [236009857](https://github.com/dotnet/machinelearning/pull/1618#discussion_r236009857) [](ancestors = 236009857,236003674,235555841,235555650)

**Comment by codemzs** [comment]
Thanks, Pete. The two tests that were testing batch prediction are now passing. 

---
In reply to: [441307172](https://github.com/dotnet/machinelearning/pull/1618#issuecomment-441307172) [](ancestors = 441307172)

File: src/Microsoft.ML.Data/DataLoadSave/TransformerChain.cs (hunk 20 above)
Comment (codemzs) [reply src/Microsoft.ML.Data/DataLoadSave/TransformerChain.cs]: I created this interface so that I can access transformers and scope from an ITransfomer reference because casting ITransfomer to TransformerChain<ITransformer> is messy.

---
In reply to: [236007746](https://github.com/dotnet/machinelearning/pull/1618#discussion_r236007746) [](ancestors = 236007746,236007659)

File: src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs (hunk 23 above)
[the remark this answers is not in the document]
Comment (codemzs) [reply src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs]: Done.

File: src/Microsoft.ML.Api/PredictionFunction.cs (hunk 18 above)
Comment (codemzs) [reply src/Microsoft.ML.Api/PredictionFunction.cs]: I'm with you 100% on this. I have the same thought when I was making this change.

---
In reply to: [236007234](https://github.com/dotnet/machinelearning/pull/1618#discussion_r236007234) [](ancestors = 236007234)

File: src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs (hunk 16 above)
Comment (codemzs) [reply src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs]: i'll refactor this out and make it internal

---
In reply to: [235554740](https://github.com/dotnet/machinelearning/pull/1618#discussion_r235554740) [](ancestors = 235554740)

File: src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs (hunk 22 above)
Comment (codemzs) [reply src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs]: Sounds good.

---
In reply to: [236008749](https://github.com/dotnet/machinelearning/pull/1618#discussion_r236008749) [](ancestors = 236008749)

File: src/Microsoft.ML.TimeSeries/PredictionEngine.cs [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+    /// </summary>
+    /// <typeparam name="TSrc">The user-defined type that holds the example.</typeparam>
+    /// <typeparam name="TDst">The user-defined type that holds the prediction.</typeparam>
+    public sealed class TimeSeriesPredictionEngine<TSrc, TDst> : PredictionEngineBase<TSrc, TDst>
+        where TSrc : class
+        where TDst : class, new()
+    {
+        private ValuePinger[][] _pingers;
+        private int _rowPosition;
+        public ITransformer CheckPoint => Transformer;
```
[the remark this answers is not in the document]
Comment (codemzs) [reply src/Microsoft.ML.TimeSeries/PredictionEngine.cs]: I'm confused but please help me understand your concern here. Isn't that what I'm doing below? I have changed the API for checkpointing where the user passes a modelpath and prediction engine writes the model with updated state there. 

---
In reply to: [236009638](https://github.com/dotnet/machinelearning/pull/1618#discussion_r236009638) [](ancestors = 236009638)
